src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,360 @@
+
+    fallback_to_stale_replicas = settings && settings->fallback_to_stale_replicas_for_distributed_queries;
+    entries_count = 0;
+    usable_count = 0;
+    failed_pools_count = 0;
+}
+
+HedgedConnectionsFactory::~HedgedConnectionsFactory()
+{
+    pool->updateSharedError(shuffled_pools);
+}
+
+std::vector<Connection *> HedgedConnectionsFactory::getManyConnections(PoolMode pool_mode)
+{
+    size_t min_entries = (settings && settings->skip_unavailable_shards) ? 0 : 1;
+
+    size_t max_entries;
+    if (pool_mode == PoolMode::GET_ALL)
+    {
+        min_entries = shuffled_pools.size();
+        max_entries = shuffled_pools.size();
+    }
+    else if (pool_mode == PoolMode::GET_ONE)
+        max_entries = 1;
+    else if (pool_mode == PoolMode::GET_MANY)
+        max_entries = settings ? size_t(settings->max_parallel_replicas) : 1;
+    else
+        throw DB::Exception("Unknown pool allocation mode", DB::ErrorCodes::LOGICAL_ERROR);
+
+    std::vector<Connection *> connections;
+    connections.reserve(max_entries);
+    Connection * connection = nullptr;
+
+    /// Try to start establishing connections with max_entries replicas.
+    for (size_t i = 0; i != max_entries; ++i)
+    {
+        int index = startEstablishingNewConnection(connection);
+        if (index == -1)
+            break;
+        if (replicas[index].is_ready)
+            connections.push_back(connection);
+    }
+
+    /// Process connections until we get enough READY connections
+    /// (work asynchronously with all connections we started).
+    while (connections.size() < max_entries)
+    {
+        auto state = getNextConnection(false, true, connection);
+        if (state == State::READY)
+            connections.push_back(connection);
+        else if (state == State::CANNOT_CHOOSE)
+        {
+            if (connections.size() >= min_entries)
+                break;
+
+            /// Determine the reason of not enough replicas.
+            if (!fallback_to_stale_replicas && usable_count >= min_entries)
+                throw DB::Exception(
+                    "Could not find enough connections to up-to-date replicas. Got: " + std::to_string(connections.size())
+                    + ", needed: " + std::to_string(min_entries),
+                    DB::ErrorCodes::ALL_REPLICAS_ARE_STALE);
+
+            throw DB::NetException(
+                "All connection tries failed. Log: \n\n" + fail_messages + "\n",
+                DB::ErrorCodes::ALL_CONNECTION_TRIES_FAILED);
+        }
+    }
+
+    return connections;
+}
+
+HedgedConnectionsFactory::State HedgedConnectionsFactory::getNextConnection(bool start_new_connection, bool blocking, Connection *& connection_out)
+{
+    if (start_new_connection)
+    {
+        int index = startEstablishingNewConnection(connection_out);
+        if (index != -1 && replicas[index].is_ready)
+            return State::READY;
+    }
```

Comment:
I would recommend to extract this to separate method with name `startNewConnection`. And rename current method to something like `waitForReadyConnection` (which will also check that there is something to wait).
And remove `start_new_connection`.
Such boolean flags significantly decrease code readability.